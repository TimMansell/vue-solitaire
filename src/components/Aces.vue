<template>
  <div class="aces">
    <div
      class="ace"
      v-for="(ace, index) in aces"
      :key="index"
      @click="moveCardToAce"
      data-test="ace">
      <span v-if="!ace">A</span>
      <Card
        v-for="(card, index2) in ace"
        :key="index2"
        :value="card.value"
        :suit="card.suit"
        :order="card.order"
        :position="card.position"
        :visible="card.visible"
        :clickable="false" />
    </div>
  </div>
</template>

<script>
import Card from '@/components/Card.vue';

export default {
  name: 'Aces',
  components: {
    Card,
  },
  data() {
    return {
      aces: this.$store.getters.aces,
    };
  },
  methods: {
    moveCardToAce() {
      this.$store.dispatch('moveCardToAce');
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
.aces {
  display: flex;
  height: 52px;
  margin-bottom: 1rem;
  padding: .5rem;

  @media (min-width: 900px) {
    height: 100px;
  }
}

.ace {
  position: relative;
  border: 1px solid black;
  margin-right: 1rem;
  width: 100px;
}

.card {
  position: absolute;

  &:nth-of-type(n+2) {
    margin-top: 0;
    z-index: 1;
  }
}
</style>
