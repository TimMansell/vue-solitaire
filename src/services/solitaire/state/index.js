const setState = (state, newState) => ({ ...state, ...newState });

export default setState;
