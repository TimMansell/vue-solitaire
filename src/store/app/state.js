// Return state as function so we can reset it.
const state = () => ({
  game: {
    moves: [],
    time: 0,
  },
  gameOutcome: {
    hasGameWon: false,
    hasGameLost: false,
  },
  isGameLoading: true,
  isGamePaused: false,
  isOverlayVisible: true,
  versionMatch: true,
});

export default state;
