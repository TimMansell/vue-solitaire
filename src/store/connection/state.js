const state = () => ({
  isConnecting: false,
  hasConnected: false,
  isOnline: false,
});

export default state;
