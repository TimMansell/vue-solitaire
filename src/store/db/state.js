// Return state as function so we can reset it.
const state = () => ({
  userStats: {
    totalGames: null,
  },
  game: {
    id: null,
    moves: 0,
  },
});

export default state;
