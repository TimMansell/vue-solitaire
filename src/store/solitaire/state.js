// Return state as function so we can reset it.
const state = () => ({
  board: {
    cards: [],
    foundation: [],
  },
  selectedCardId: null,
  isGameWon: false,
  hasMoves: true,
});

export default state;
