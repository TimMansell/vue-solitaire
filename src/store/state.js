const state = {
  cards: {
    values: ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'],
    suits: ['c', 's', 'h', 'd'],
  },
  shuffledCards: [],
  board: {
    cards: [],
    foundation: [],
  },
  rules: {
    columns: [7, 7, 7, 7, 6, 6, 6, 6],
    foundationColumns: [1, 1, 1, 1],
  },
  selectedCards: [],
};

export default state;
