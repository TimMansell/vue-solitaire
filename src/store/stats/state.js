// Return state as function so we can reset it.
const state = () => ({
  globalStats: [],
  userStats: [],
  playerCount: 0,
  onlinePlayerCount: 0,
  userGameCount: 0,
  globalGameCount: 0,
  leaderboards: [],
});

export default state;
