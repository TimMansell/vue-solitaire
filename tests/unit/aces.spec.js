import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Aces from '@/components/Aces.vue';
import state from '@/store/state';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Aces.vue', () => {
  it('matches snapshot', () => {
    const getters = {
      aces: () => ({
        c: [{
          suit: 'c',
          value: 'A',
        }],
        h: [{
          suit: 'h',
          value: 'A',
        }],
        d: [{
          suit: 'd',
          value: 'A',
        }],
        s: [{
          suit: 's',
          value: 'A',
        }],
      }),
    };

    const store = new Vuex.Store({
      getters,
    });

    const wrapper = shallowMount(Aces, { store, localVue });

    expect(wrapper).toMatchSnapshot();
  });

  it('calls store action "moveCardToAce" when clicked', () => {
    const getters = {
      aces: () => state.board.aces,
    };

    const actions = {
      moveCardToAce: jest.fn(),
    };

    const store = new Vuex.Store({
      getters,
      actions,
    });

    const wrapper = shallowMount(Aces, { store, localVue });

    wrapper.find('[data-test="ace"]').trigger('click');

    expect(actions.moveCardToAce).toHaveBeenCalled();
  });
});
