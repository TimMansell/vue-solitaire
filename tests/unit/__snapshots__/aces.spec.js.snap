// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aces.vue matches snapshot 1`] = `
<div data-test="ace" class="aces">
  <div data-test="ace-c" class="ace">
    <!---->
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
  </div>
  <div data-test="ace-h" class="ace">
    <!---->
    <card-stub value="A" suit="h" order="0" position=""></card-stub>
  </div>
  <div data-test="ace-d" class="ace">
    <!---->
    <card-stub value="A" suit="d" order="0" position=""></card-stub>
  </div>
  <div data-test="ace-s" class="ace">
    <!---->
    <card-stub value="A" suit="s" order="0" position=""></card-stub>
  </div>
</div>
`;
